testplan/common/utils/graph.py

```diff
@@ -0,0 +1,190 @@
+from collections import defaultdict
+from copy import copy
+from dataclasses import dataclass
+from typing import Dict, Generic, List, Tuple, TypeVar
+
+try:
+    from typing import Self  # starting from py311
+except ImportError:
+    from typing_extensions import Self
+
+from memoization import cached
+
+T = TypeVar("T")  # vertex rep
+U = TypeVar("U")  # actual vertex
+V = TypeVar("V")  # actual edge
+
+GraphComponent = List[T]  # a component is a group of vertices
+
+
+@dataclass
+class DirectedGraph(Generic[T, U, V]):
+    """adjacency matrix for directed graph"""
+
+    vertices: Dict[T, U]
+    edges: Dict[T, Dict[T, V]]
+    indegrees: Dict[T, int]
+    outdegrees: Dict[T, int]
+
+    @classmethod
+    def new(cls) -> Self:
```

Comment:
Why do we need this classmethod? Is `field` with `default_factory` limiting somehow here?

Reply:
tbh it's personal preference (

Reply:
I have no strong preference, so we can go this way, I leave it to other reviewers to weigh in if they would like to.